operators/src/util/sunpos.rs

```diff
@@ -0,0 +1,138 @@
+    greenwich_mean_sidereal_time: f64,
+    right_ascension: f64,
+    declination: f64,
+}
+
+impl SunPos {
+    pub fn solar_azimuth_zenith(&self, lat: f64, lon: f64) -> (f64, f64) {
+        let local_mean_sidereal_time =
+            (self.greenwich_mean_sidereal_time * 15.0 + lon).to_radians();
+        let hour_angle = local_mean_sidereal_time - self.right_ascension;
+
+        let mut zenith = (lat.to_radians().cos() * hour_angle.cos() * self.declination.cos()
+            + self.declination.sin() * lat.to_radians().sin())
+        .acos();
+
+        let parallax = (EARTH_MEAN_RADIUS / ASTRONOMICAL_UNIT) * zenith.sin();
+        zenith += parallax;
+
+        let dx = self.declination.tan() * lat.to_radians().cos()
+            - lat.to_radians().sin() * hour_angle.cos();
+
+        let dy = -hour_angle.sin();
+
+        let mut azimuth = dy.atan2(dx);
+
+        if azimuth < 0.0 {
+            azimuth += std::f64::consts::TAU;
+        }
+
+        (azimuth.to_degrees(), zenith.to_degrees())
+    }
+
+    /// Compute the sunpos at the given date.
+    pub fn new<T: TimeZone>(timestamp: &DateTime<T>) -> SunPos {
+        let (decimal_hours, elapsed_julian_days) = Self::elapsed_julian_days(timestamp);
+        let (ecliptic_longitude, ecliptic_obliquity) =
+            Self::eliptic_coordinates(elapsed_julian_days);
+
+        let (right_ascension, declination) =
+            Self::celestial_coordinates(ecliptic_longitude, ecliptic_obliquity);
+
+        let greenwich_mean_sidereal_time =
+            6.697_424_324_2 + 0.065_709_828_3 * elapsed_julian_days + decimal_hours;
+
+        SunPos {
+            greenwich_mean_sidereal_time,
+            right_ascension,
+            declination,
+        }
+    }
+
+    /// Calculates difference in days between the current Julian Day
+    /// and JD 2451545.0, which is noon 1 January 2000 Universal Time
+    fn elapsed_julian_days<T: TimeZone>(timestamp: &DateTime<T>) -> (f64, f64) {
+        let year = i64::from(timestamp.year());
+        let month = i64::from(timestamp.month());
+        let day = i64::from(timestamp.day());
+
+        let hour = f64::from(timestamp.hour());
+        let minute = f64::from(timestamp.minute());
+        let second = f64::from(timestamp.second());
+
+        let decimal_hours = hour + (minute + second / 60.0) / 60.0;
+        let aux1 = (month - 14) / 12;
+        let aux2 = (1461 * (year + 4800 + aux1)) / 4 + (367 * (month - 2 - 12 * aux1)) / 12
+            - (3 * ((year + 4900 + aux1) / 100)) / 4
+            + day
+            - 32075;
+        let julian_date = aux2 as f64 - 0.5 + decimal_hours / 24.0;
+        (decimal_hours, julian_date - 2_451_545.0)
+    }
+
+    /// Calculate ecliptic coordinates (ecliptic longitude and obliquity of the
+    /// ecliptic in radians but without limiting the angle to be less than 2*Pi
+    /// (i.e., the result may be greater than 2*Pi)
+    fn eliptic_coordinates(elapsed_julian_days: f64) -> (f64, f64) {
```

Comment:
```suggestion
    /// Calculate elliptic coordinates (elliptic longitude and obliquity of the
    /// elliptic in radians but without limiting the angle to be less than 2*Pi
    /// (i.e., the result may be greater than 2*Pi)
    fn elliptic_coordinates(elapsed_julian_days: f64) -> (f64, f64) {
```
if you mean *elliptic*

Reply:
I meant _ecliptic_ :D Will fix this.

Reply:
21be41da5f29bed84c76fa94581c5512e0eb0c8e